### PR TITLE
Updates to newest available Docker

### PR DESCRIPTION
**Pull request by adriancole**
Currently, we are adding syntax overrides in Dockerfile to avoid old bugs
in Docker 17. This moves up to latest Docker available in a Java image, which
is only one version behind CirclCI's normal Docker image.

See https://discuss.circleci.com/t/docker-v19-03-12-now-available-in-remote-docker/36714

### CODE REVIEW

File: .circleci/config.yml [hunk 1]
```diff
@@ -3,7 +3,9 @@ version: 2.1
 executors:
   gradle_docker:
     docker:
-      - image: cimg/openjdk:11.0
+      # This image isn't used to build Java project, rather to invoke Docker with Gradle.
+      # Using latest JDK image will give more recent Docker, albeit slightly behind cimg/base
```
Comment (aaron-steinfeld) [inline .circleci/config.yml:7]: Since we're using a remote docker anyway, is this change related to upgrading docker? Does the client run locally by the build need to match the remote?

File: .circleci/config.yml [hunk 2]
```diff
@@ -30,7 +32,10 @@ commands:
       - restore_cache:
           keys: # Use gradle version for cache, so it doesn't download every time
             - v1-dependencies-{{ checksum "./gradle/wrapper/gradle-wrapper.properties" }}
-      - setup_remote_docker
+      # Update this according to the output of the image used to override the default of Docker 17
+      #   Ex. docker run -it --rm cimg/openjdk:14.0.2 docker -v
+      - setup_remote_docker:
+          version: 19.03.11
```
Comment (aaron-steinfeld) [inline .circleci/config.yml:38]: `19.03.12` is latest available https://circleci.com/docs/2.0/building-docker-images/#docker-version

Comment (adriancole) [reply .circleci/config.yml:38]: that version is on the cimg/base image, not the java one, hence the instructions mentioning check the image that we are using. I think most are using cimg/base to run docker commands, but we can't as we invoke docker with gradle.

File: .circleci/config.yml (hunk 1 above)
Comment (adriancole) [reply .circleci/config.yml:7]: I'm going to be swimming in pull requests, and it is odd to use old JDKs especially to run gradle docker commands. If possible, I'd like to update this, instead of another few dozen PRs later to do the same.

Comment (adriancole) [reply .circleci/config.yml]: changed the comment, but I'd like to leave the version update so that we have less work later. there are many many many versions behind due to a culture of not updating, which is a problem we need to address.

Comment (aaron-steinfeld) [reply .circleci/config.yml]: My hesitation on 14 is mainly based on it going EOL in three weeks (I still think java's release/support model is ridiculous...). If we're going to be updating to 15 at that time, it doesn't seem super beneficial to update to 14 now, but maybe it'll flush out some issues so 👍

File: .circleci/config.yml (hunk 2 above)
Comment (aaron-steinfeld) [reply .circleci/config.yml]: the `cimg/base:stable` got me there (because of the java requirement), but the change looks good.

File: .circleci/config.yml (hunk 1 above)
Comment (adriancole) [reply .circleci/config.yml]: Luckily, it doesn't have a built-in bladerunneresque expiry either! I don't think anyone will expect people to switch to 15 the day it is released. In fact, a lot of projects have a matrix of many JDKs. In practice, they linger a bit and I think that's acceptable.

Indeed, the highest value is flushing out issues so we don't have a massive pile of them later. If we find things we'll know as they occur when the JDK is released.

**Comment by adriancole** [comment]
Sorry I didn't make the tracking issue earlier, it makes sense to have done that first. here it is https://github.com/hypertrace/hypertrace/issues/22
